### ISSUE TITLE
파이프라인 바텀업 검증: SMPL 피팅 → VTP 시각화 정확도 수정 (#8)

### ISSUE THREAD

**Issue by sigongjoa** [sigongjoa/Ballet-Sorcery#8]
## 문제 요약

이슈 #7에서 구현된 SMPL 피팅 → 가상 마커 → TRC → OpenSim VTP 파이프라인이
**수치 지표(각도 오차 12.7°)는 개선됐지만, 실제 시각 결과는 명백히 틀림**.

### 관찰된 증상 (Frame 0600 기준)
- 원본: 양 다리 벌림 (스탠스 자세)
- VTP 결과: 두 다리가 골반 한쪽에 몰림 / 겹침
- 리포트 텍스트("정확히 반영됨")와 시각 결과 불일치

### 근본 원인 가설
```
각 단계의 정확도가 한 번도 개별 검증되지 않음:
MediaPipe(33pt) → SMPL 피팅 → 가상 마커(54pt) → FK 각도 → VTP 렌더링
       ?              ?              ?               ?         ?
```
단순 코드 버그가 아닌 **알고리즘 설계/정확도 누적 붕괴** 문제로 판단.

---

### 표준 검증 메트릭 도입
| 메트릭 | 설명 | 목표 |
|---|---|---|
| **재투영 오차** | SMPL joint → 2D 재투영 → 원본 keypoint 비교 | < 10px |
| **MPJPE** | 관절 위치 평균 오차 (mm) | < 80mm |
| **좌우 대칭성** | L/R 마커 x좌표 부호 일치 여부 | 100% |
| **관절 각도 오차** | 알려진 포즈 vs 예측 | < 5° |

#### Step 1: SMPL T-pose 기준 테스트 (합성 데이터)
```python
# body_pose=zeros → T-pose 생성 → joint 위치 대칭성 확인
model = smplx.create(model_dir, model_type='smpl', ...)
out = model(body_pose=torch.zeros(1, 69))
# L_Hip.x < 0, R_Hip.x > 0 인지 확인
```
- [ ] `tools/verify_tpose_symmetry.py` 작성

#### Step 2: SMPL 피팅 재투영 오차 시각화
- SMPL joint 위치 → 카메라 투영 → 원본 영상 2D 위에 오버레이
- 관절이 원본 랜드마크와 정렬되는지 프레임별 확인
- [ ] `tools/verify_smpl_reprojection.py` 작성

#### Step 3: 가상 마커 좌우 대칭성 검증
- 54개 마커 중 L/R 쌍에 대해 x좌표 부호 검증
- `ACROMION_L.x < 0` AND `ACROMION_R.x > 0` 등
- [ ] `tools/verify_marker_symmetry.py` 작성

#### Step 4: FK 단일 관절 각도 수치 검증
- Frame 0600 (다리 벌림 자세)에서 `hip_adduction_r` 각도값 출력
- 기대값: 약 20~40° (실제 스탠스 각도)
- [ ] `tools/verify_fk_single_joint.py` 작성

#### Step 5: VTP 카메라 정면 뷰 + zoom-to-fit
- pelvis FK 위치 기준 동적 카메라 배치
- 전신이 화면 중앙 80%에 들어오도록
- [ ] `src/anatomy_overlay_viewer.py` prerender_vtp_frames() 수정

### 검증 데이터
- **합성**: SMPL T-pose (파라미터 zeros) → 정답이 알려진 상태
- **실제**: IMG_2633.MOV Frame 0600 (다리 벌림) — 시각적으로 명확한 포즈
- **참고**: AMASS 데이터셋 (추후) — 344명 SMPL 파라미터 정답 포함

---

## 완료 기준

- [ ] T-pose SMPL joint 대칭성 테스트 통과
- [ ] Frame 0600 재투영 오차 < 30px (현실적 목표)
- [ ] L/R 마커 좌우 부호 100% 일치
- [ ] Frame 0600 VTP에서 두 다리가 시각적으로 분리되어 보임
- [ ] 파이프라인 각 단계 검증 리포트 생성

## 참고 자료
- [Pose2Sim Accuracy Paper](https://www.mdpi.com/1424-8220/22/7/2712) — 관절 각도 MAE 0.35°~1.6° 달성 방법
- [VIBE SMPLify losses](https://github.com/mkocabas/VIBE/blob/master/lib/smplify/losses.py) — 재투영 오차 계산 구현
- [AMASS Dataset](https://amass.is.tue.mpg.de/) — SMPL 정답 파라미터 포함 데이터셋

**Comment by sigongjoa** [comment]
## 🔄 Pose Prior 전략 재검토 — Option C: VPoser 채택

### GMM Prior (Option A) 재고

직전 코멘트에서 `gmm_08.pkl` 활성화를 제안했는데, **이건 발레에 역효과다.**

`gmm_08.pkl`의 학습 데이터:
- CMU Mocap 기반 일상 동작 (걷기, 뛰기, 앉기)
- 고관절 굴곡 최대 ~90°, 턴아웃 ~20° 수준

발레 동작과의 괴리:
- 아라베스크: 고관절 굴곡 150°+
- à la seconde: 고관절 외전 90°+
- 다리 올리기 (Frame 1096): GMM 분포에서 **이상치(outlier)**

```
GMM loss가 증가하면 optimizer가 포즈를 "일반적인" 방향으로 당긴다.
발레 극단 포즈에서 GMM prior가 강하게 작동 → data term을 이기면 정반대 효과.
```

---

### Option C: VPoser (VAE-based Pose Prior)

**왜 VPoser인가:**

| 항목 | GMM (Option A) | VPoser (Option C) |
|------|---------------|-------------------|
| 모델 구조 | 8-component GMM | VAE (Variational AutoEncoder) |
| 학습 데이터 | ~1700 시퀀스 CMU Mocap | AMASS 40시간+ (발레 포함) |
| 극단 포즈 처리 | ❌ 이상치로 강하게 패널티 | ✅ 잠재공간에서 연속적 표현 |
| SMPL-X 호환 | ⚠️ SMPL body_pose만 | ✅ SMPL-X 공식 지원 |
| 우리 파이프라인 적합성 | 낮음 | **높음** |

VPoser는 SMPL-X 팀(Max Planck)이 SMPL-X와 함께 배포하며,
AMASS 데이터셋(발레, 스포츠, 일상 동작 포함)으로 학습되어 있다.

---

### 취득 방법

**Step 1: 공식 파일 다운로드**
```
https://smpl-x.is.tue.mpg.de/ (free academic registration)
VPoser v2.0 → V02_05 폴더 (snapshots/V02_05_epoch=08_val_loss=0.03.ckpt)
```

**Step 2: 의존성 설치**
```bash
pip install human_body_prior  # VPoser 공식 패키지 (PyTorch 기반)
```

**Step 3: 파이프라인 통합**
```python
from human_body_prior.tools.model_loader import load_model
from human_body_prior.models.vposer_model import VPoser

vposer, _ = load_model('data/models/vposer_v2_0',
                        model_code=VPoser,
                        remove_words_in_model_weights='vp_model.')

# smplx_engine.py 수정: body_pose를 직접 최적화 대신
# latent z를 최적화, VPoser decoder로 body_pose 생성
z = torch.zeros(1, 32, requires_grad=True)  # 32-dim latent
body_pose = vposer.decode(z, output_type='aa').view(1, 63)  # 21 joints × 3
```

---

### 발레 데이터 파인튜닝 전략

기본 VPoser v2.0으로도 개선되지만, 발레 특화를 위해 추가 파인튜닝이 가능하다.

**데이터 준비:**
```
우리 파이프라인: MediaPipe → SMPL fitting → body_pose 파라미터 수집
IMG_2633.MOV 2280 프레임 → body_pose 2280개 (발레 plié, arabesque 등)
```

**파인튜닝 방법:**
1. AMASS pretrained VPoser를 base로 사용
2. 우리 발레 body_pose 데이터로 few-shot fine-tuning
3. KL divergence loss로 기존 latent space 유지하면서 발레 포즈 추가

> **중요**: 파인튜닝 전에도 기본 VPoser v2.0 적용만으로도
> 현재 GMM-없는 상태보다 개선될 것으로 예상.
> 파인튜닝은 Phase 2로 분리.

---

### 검증 프로세스

**기준선 (현재):**
| 자세 | MPJPE | 재투영 오차 |
|------|-------|------------|
| 일반 서기/보행 | 5.87 cm | ~33 px |
| 발레 극단 포즈 | ~25 cm | 73.7 px |

**VPoser 적용 후 측정:**
```python
# tools/e2e_smpl_error.py 에서 동일 5 프레임 재측정
# 비교 프레임: 200(일반), 500(일반), 800(일반), 1400(일반), 1096(발레 다리 올리기)
```

**통과 기준:**
- [ ] 일반 포즈 MPJPE: 5.87 cm → < 4 cm (15% 이상 개선)
- [ ] 발레 극단 포즈 MPJPE: ~25 cm → < 10 cm (60% 이상 개선)
- [ ] 재투영 오차 Frame 600: 73.7 px → < 40 px

**시각적 검증:**
- 동일 `tools/e2e_reprojection_overlay.py`로 Frame 100/300/600 재생성
- 발레 포즈에서 SMPL joint가 MediaPipe landmark에 가까워지는지 확인

---

### 구현 계획

```
Phase 1 (현재): VPoser v2.0 취득 + smplx_engine.py latent-space 최적화 통합
  → 측정: 동일 5 프레임 MPJPE 비교 (before/after)

Phase 2 (추후): 발레 body_pose 데이터 수집 → VPoser 파인튜닝
  → 측정: arabesque, grand battement 등 극단 포즈 전용 벤치마크
```

- [ ] `data/models/vposer_v2_0/` 폴더에 VPoser 모델 배치
- [ ] `src/smplx_engine.py` — z latent 최적화 방식으로 전환
- [ ] `tools/e2e_smpl_error.py` 재실행 → before/after 비교 리포트

**Comment by sigongjoa** [comment]
## ✅ Phase 1 완료 — 2-stage 최적화 + 발레 각도 제한 Prior

### 구현 내용

`smplx_engine.py` 전면 개선:

1. **2-stage SMPLify 최적화**
   - Stage 1 (30 iter, lr=0.05): `global_orient + transl` 만 — 전체 위치/방향 먼저 정렬
   - Stage 2 (170 iter, lr=0.01→0.0001): 전체 파라미터 — CosineAnnealingLR 적용

2. **발레 허용 각도 제한 Prior** (GMM 대신)
   ```
   고관절(L/R):   3.0 rad (172°) — 아라베스크, 그랑 바트망 허용
   무릎:          2.9 rad (166°) — 플리에 허용
   어깨:          3.0 rad (172°) — 포르드브라 허용
   척추:          1.2 rad (69°)  — 물리적 범위 제한
   손목:          1.0 rad (57°)  — 물리적 범위 제한
   ```
   → 발레 극단 포즈는 패널티 없이 통과, 물리 불가능한 포즈만 억제

3. **VPoser 자동 활성화 스텁**
   - `data/models/vposer_v1_0/` 존재 시 자동 로드
   - z latent (32차원) 최적화로 전환

### Phase 1 검증 결과

| 지표 | Before (50iter, L2) | After (200iter, 2-stage) | 목표 | 결과 |
|------|--------------------|-----------------------|------|------|
| Mean MPJPE | 9.83 cm | **5.13 cm** | — | -48% |
| 일반 포즈 MPJPE | 5.87 cm | **3.75 cm** | < 4 cm | ✅ |
| 극단 포즈 MPJPE (Frame 1096) | ~25 cm | **10.62 cm** | < 12 cm | ✅ |
| 재투영 Frame 600 | 73.7 px | **50.4 px** | < 50 px | ⚠️ 근접 |
| 재투영 Frame 100/300 | ~33-36 px | ~33-34 px | — | 유지 |

**테스트**: 30 passed, 3 skipped ✅

---

### 다음: Phase 2 — Ballet-GMM 구축

Frame 600 재투영이 50.4 px (목표 50 px 에서 0.4 px 미달). VPoser 없이 이 이상 개선하려면 ballet-specific prior가 필요하다.

**Phase 2 실행 계획:**

```
① 발레 영상 데이터 수집
   - IMG_2633.MOV 2280 프레임 (현재 보유)
   - AIST Dance Database (공개)
   - 추가 발레 수업 영상

② body_pose 벡터 추출
   현재 200-iter 엔진으로 전체 2280 프레임 피팅
   → ballet_poses.npy  shape=(N, 69)

③ sklearn GMM 학습
   n_components=16, covariance_type='full'
   → data/models/smpl/ballet_gmm.pkl

④ smplx_engine.py에 ballet_gmm 통합
   → 재측정으로 목표 달성 여부 확인
```

VPoser v1.0은 https://smpl-x.is.tue.mpg.de 에서 academic 등록 후 다운로드 가능.
`data/models/vposer_v1_0/` 에 배치하면 engine이 자동 활성화.

**Comment by sigongjoa** [comment]
## 💡 GMM 사용 방법 재설계: 손실 함수 → 초기값 warm-start

### 기존 접근의 근본 문제

GMM/VPoser를 손실 함수로 쓰면 필연적으로 충돌이 발생한다:

```
T-pose(zeros)에서 시작
    ↓
GMM: "이 포즈는 발레 분포에서 멀다 → body_pose를 분포 중심으로 당겨라"
Data: "이 관절은 MediaPipe 위치에 맞춰라"
    ↓
두 gradient가 서로 다른 방향을 가리킴 → 최적화 발산
```

GMM λ를 아무리 조절해도 T-pose 초기화 + 손실 충돌 구조는 변하지 않음.

---

### 올바른 GMM 사용법: Warm-Start Initialization

동영상(2280 프레임)에서 추출한 ballet_gmm은 **이 영상의 포즈 분포**를 담고 있다.

```
GMM 가중 평균 = 영상에서 가장 자주 나타나는 발레 포즈

T-pose(zeros) 초기화:  body_pose = [0, 0, ..., 0]   (69차원)
GMM warm-start 초기화: body_pose = Σ_k w_k * μ_k     (발레 포즈 평균)
```

최적화가 T-pose에서 발레 포즈까지 이동할 필요 없이, 이미 근처에서 시작.
손실 함수에서 완전히 제거 → gradient 충돌 없음.

---

### 구현 계획

```python
# Stage 2 시작 시 body_pose 초기화
if self.ballet_gmm is not None:
    # GMM 가중 평균으로 warm-start
    body_pose_init = (weights[:, None] * means).sum(axis=0)  # (69,)
    body_pose = torch.tensor(body_pose_init, device=device).unsqueeze(0)
    body_pose.requires_grad_(True)
else:
    body_pose = torch.zeros(1, 69, ...)  # 기존

# 손실 함수: GMM 제거, 각도 제한만 유지
loss_pose = (bp ** 2).sum() * 0.005
loss_pose += angle_limit_loss(bp) * 2.0
```

추가 개선: **프레임 간 temporal warm-start**
- 이전 프레임의 body_pose → 다음 프레임 초기값
- 동영상에서 연속 프레임은 포즈 변화가 작음 → 수렴 빠름

---

### 기대 효과

| | 현재 (T-pose 시작) | GMM warm-start |
|---|---|---|
| Stage 2 시작점 | 발레 포즈에서 멀리 | 발레 포즈 근처 |
| 필요 iteration | 170 | 50~100 (추정) |
| 극단 포즈 MPJPE | 10.62 cm | 개선 기대 |
| gradient 충돌 | 없음 (GMM 제거) | 없음 |

다음 구현: `smplx_engine.py` warm-start 적용 → 동일 벤치마크 재측정
